### PR TITLE
Software version paralel with full functional firemware.

### PR DESCRIPTION
**Pull request by qiguaiderenmen**
Commenets & code refine will be done later.

### CODE REVIEW

File: lib/code/AmpIO.cpp [hunk 1]
```diff
@@ -1502,8 +1502,10 @@ bool AmpIO::DallasWaitIdle()
 {
     int i;
     AmpIO_UInt32 status;
-    // Wait up to 500 msec. Based on measurements, approximate wait time is 250-300 msec.
-    for (i = 0; i < 500; i++) {
+    // Wait up to 2000 msec. 
+    // Based on measurements, direct 1-wire interface's approximate wait time is 250-300 msec.
+    // For DS2480B serial method, the approximate wait time is about 4-5 sec.
```
Comment (pkazanzides) [inline lib/code/AmpIO.cpp]: If the DS2480B wait time is 4-5 seconds, why are you only waiting 2 seconds?

File: lib/code/AmpIO.cpp [hunk 2]
```diff
@@ -1513,22 +1515,22 @@ bool AmpIO::DallasWaitIdle()
...
-    if (useDS2480B) ctrl |= 4;
     if (!DallasWriteControl(ctrl)) return false;
     if (!DallasWaitIdle()) return false;
     if (!DallasReadStatus(status)) return false;
-    // Check family_code, dout_cfg_bidir, ds_reset, and ds_enable
-    if ((status & 0xFF00000F) != 0x0B00000B) return false;
+    
+    // Automatically detect interface in use
+    useDS2480B = (status & 0x00008000) == 0x00008000; 
+    if (useDS2480B) ctrl |= 4;  // corresponds to reg_wdata[2] in firmware
```
Comment (pkazanzides) [inline lib/code/AmpIO.cpp]: With the auto-detect feature, I don't think reg_wdata[2] is needed anymore

File: lib/code/AmpIO.cpp [hunk 3]
```diff
@@ -1538,13 +1540,19 @@ bool AmpIO::DallasReadMemory(unsigned short addr, unsigned char *data, unsigned
     nbytes -= nb;
     // Read additional blocks of data if necessary
     while (nbytes > 0) {
-        if (!DallasWriteControl(3)) return false;
+        ctrl |= 1;
```
Comment (pkazanzides) [inline lib/code/AmpIO.cpp]: Why change this?

File: lib/code/AmpIO.cpp [hunk 4]
```diff
@@ -1538,13 +1540,19 @@ bool AmpIO::DallasReadMemory(unsigned short addr, unsigned char *data, unsigned
...
+        if (!DallasWriteControl(ctrl)) return false;
         if (!DallasWaitIdle()) return false;
         nb = (nbytes>256) ? 256 : nbytes;
         if (!port->ReadBlock(BoardId, address, reinterpret_cast<quadlet_t *>(ptr), nb)) return false;
         ptr += nb;
         nbytes -= nb;
     }
+    // End all blocks reading for DS2480B interface
+    if (nbytes <= 0 && useDS2480B) {
+        if (!DallasWriteControl(13)) return false;
```
Comment (pkazanzides) [inline lib/code/AmpIO.cpp]: Maybe use hex so the bit settings are more obvious?

File: tests/instrument.cpp [hunk 5]
```diff
@@ -141,28 +134,6 @@ int main(int argc, char** argv)
     // No longer need these
     Port->RemoveBoard(board);
     delete Port;
-    if ((status & 0x00000001) != 0x00000001) {
```
Comment (pkazanzides) [inline tests/instrument.cpp]: Most of this code can be restored now.

**Review by pkazanzides — CHANGES_REQUESTED** [review]
Some cleanup/revision still needed.

File: lib/code/AmpIO.cpp (hunk 1 above)
Comment (qiguaiderenmen) [reply lib/code/AmpIO.cpp]: 4-5 seconds is the time in total for DS2480B. 2 seconds is for a single block read, say 256 bytes. We have 4*256 = 2048 bytes to read in total.

File: tests/instrument.cpp (hunk 5 above)
Comment (qiguaiderenmen) [reply tests/instrument.cpp]: THe interface enable thing is restored, but I don't think we still have those unexpected_idx things, maybe we should just remove them.

File: lib/code/AmpIO.cpp (hunk 4 above)
Comment (qiguaiderenmen) [reply lib/code/AmpIO.cpp]: Sure, now hex and comments added

File: lib/code/AmpIO.cpp (hunk 3 above)
Comment (qiguaiderenmen) [reply lib/code/AmpIO.cpp]: No need indeed, change back and add comments.

File: lib/code/AmpIO.cpp (hunk 2 above)
Comment (qiguaiderenmen) [reply lib/code/AmpIO.cpp]: Yes, and parallel modification made in firmware.

File: lib/code/AmpIO.cpp (hunk 1 above)
Comment (pkazanzides) [reply lib/code/AmpIO.cpp]: O.K. But fix the comment.  It seems like a bug if you say to wait up to 2000 msec but then say that the wait time is 4-5 sec.

Comment (qiguaiderenmen) [reply lib/code/AmpIO.cpp]: That's true. Hopefully now the comment looks better.

**Pull request merged**